### PR TITLE
[FLINK-26482][python] Support WindowedStream.reduce in Python DataStream API

### PR DESCRIPTION
**Pull request by Vancior**
## What is the purpose of the change

This pull request implements `WindowedStream.reduce` in Python DataStream API.

## Verifying this change

This change added tests and can be verified as follows:
  - Added integration tests in test_data_stream.py

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (Sphinx doc)

### CODE REVIEW

**Review by dianfu — COMMENTED** [review]
@Vancior Thanks a lot for the PR. I have left a few comments. Besides, we should also update the documentation which is located in https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/dev/python/datastream/operators/windows/

File: flink-python/pyflink/datastream/functions.py [hunk 1]
```diff
@@ -1004,15 +1010,37 @@ def process(self,
                 key: KEY,
                 window: W,
                 context: InternalWindowContext,
-                input_data: Iterable[IN]) -> Iterable[OUT]:
```
[the remark this answers is not in the document]
Comment (Vancior) [reply flink-python/pyflink/datastream/functions.py:1007]: This's an alignment to the java code, which is used to make a generic difference between single-value functions and iterable functions.

File: flink-python/pyflink/datastream/data_stream.py [hunk 2]
```diff
@@ -1347,8 +1353,58 @@ def allowed_lateness(self, time_ms: int):
...
+            >>> ds.key_by(lambda x: x[1]) \
+                    .window(TumblingEventTimeWindow.of(Time.seconds(5))) \
+                    .reduce(lambda a, b: a[0] + b[0], b[1])
+
+        :param reduce_function: The reduce function.
+        :param window_function: The window function.
+        :param output_type: Type information for the result type of the window function.
+        :return: The data stream that is the result of applying the reduce function to the window.
+
+        .. versionadded:: 1.16
```
Comment (dianfu) [inline flink-python/pyflink/datastream/data_stream.py:1384]:
```suggestion
        .. versionadded:: 1.16.0
```

File: flink-python/pyflink/datastream/data_stream.py [hunk 3]
```diff
@@ -1347,8 +1353,58 @@ def allowed_lateness(self, time_ms: int):
...
+        element per key is stored. Sliding time windows will aggregate on the granularity of the
+        slide interval, so a few elements are stored per key (one per slide interval). Custom
+        windows may not be able to incrementally aggregate, or may need to store extra values in an
+        aggregation tree.
+
+        Example:
+        ::
+
+            >>> ds.key_by(lambda x: x[1]) \
+                    .window(TumblingEventTimeWindow.of(Time.seconds(5))) \
```
Comment (dianfu) [inline flink-python/pyflink/datastream/data_stream.py:1376]:
```suggestion
                    .window(TumblingEventTimeWindows.of(Time.seconds(5))) \
```

**Review by dianfu — APPROVED** [review]
@Vancior Thanks for the update. LGTM.

**Pull request closed without merging**